### PR TITLE
Attempt for haptic support ( #341 )

### LINKED ISSUE #341: Add haptics functions

### PR DESCRIPTION
**Pull request by gnysek**
This commit adds basic gamepad vibration functions, using standard `gamepad_set_vibration()`.

Three options are available:
`input_vibrate(_strength, _time, _player_index = 0)` - vibrates both motors with `_strength` for `_frames`.

`input_vibrate_lr(_lstrength, _rstrength, _time, _player_index = 0)` - vibrates left and right motor with different strength.

`function input_vibrate_curve(_curve, _time, _player_index = 0)` - uses Animation Curve to vibrate. If curve have one path, then it's used for both motor, if more, then path 1 is used for left one, and path 2 for right one. Others are skipped.

In all cases, if `INPUT_TIMER_MILLISECONDS` set to true, then `_time` is rather milliseconds than frames.

---
fixes #341

### CODE REVIEW

**Comment by offalynne** [comment]
thanks for this. personally i would rather see juju’s existing implementation made ready for primetime than this be merged, as it follows existing guidelines for console platform features and presets instead of requiring curve assets.

thanks for the PR in any case !

**Comment by offalynne** [comment]
Good start, some suggestions

1. Code cleanup: match project style (tabs to spaces, conditionals, brackets, operators)
3. Force stop vibration on focus loss
4. Skip handling on unsupported devices: very few devices and platforms are supported
   - PlayStation and Xbox platform
   - Windows platform gamepad indices 0-3 only

**Comment by offalynne** [comment]
You can kill `INPUT_GAMEPADS_VIBRATE_WHEN_NO_FOCUS` -- the application should never do this :)

**Comment by offalynne** [comment]
Let's workshop this a bit more before making a PR into `dev`

**Comment by offalynne** [comment]
I'm happy with this now but anticipate Juju may have some thoughts on review

File: scripts/__input_class_gamepad/__input_class_gamepad.gml [hunk 1]
```diff
@@ -236,13 +235,25 @@ function __input_class_gamepad(_index) constructor
...
-            
-                __haptic_strength_left  = animcurve_channel_evaluate(_struct.channels[0], __haptic_step/__haptic_steps_total);
-                __haptic_strength_right = animcurve_channel_evaluate(_struct.channels[((array_length(_struct.channels) > 1)? 1 : 0)], __haptic_step/__haptic_steps_total);
+                var _animation_curve = animcurve_get(__haptic_curve);
+                var _channel_count   = array_length(_animation_curve.channels);
+                
+                if (_channel_count == 1)
+                {
+                    _channel_left  = animcurve_get_channel(_animation_curve, 0);
+                    _channel_right = animcurve_get_channel(_animation_curve, 0);
```
Comment (gnysek) [inline scripts/__input_class_gamepad/__input_class_gamepad.gml]: isn't `_channel_right = _channel_left;` better ?

File: scripts/__input_class_gamepad/__input_class_gamepad.gml [hunk 2]
```diff
@@ -236,13 +235,25 @@ function __input_class_gamepad(_index) constructor
...
+                }
+                else
+                {
+                    _channel_left  = animcurve_get_channel(__haptic_curve, "left");
+                    _channel_right = animcurve_get_channel(__haptic_curve, "right");
```
Comment (gnysek) [inline scripts/__input_class_gamepad/__input_class_gamepad.gml]: I'm not sure if forcing users to have `left` and `right` channel is good, they might have to use same curve for something else, that's why in my version I was checking channel 0 and 1 (that also required less setup from user side).

Comment (offalynne) [reply scripts/__input_class_gamepad/__input_class_gamepad.gml]: I agree the user might want the curve for something else, which is why I think using the first and second channels is too arbitrary.

Perhaps a macro for the channel names is in order. Edit: [done](https://github.com/gnysek/Input/pull/3/files)

Comment (offalynne) [reply scripts/__input_class_gamepad/__input_class_gamepad.gml:249]: https://github.com/gnysek/Input/pull/3

**Comment by offalynne** [comment]
Still a bit more naive than I'd like I think, ideally you should be able to additively send effects to either motor without interruption, but as of my last PR to your fork I think this is OK for a first pass.

Feel free to re-open once we've resolved [the channel name issue](https://github.com/JujuAdams/Input/pull/434#pullrequestreview-1067398158) to mutual satisfaction, otherwise I will

**Comment by offalynne** [comment]
See https://github.com/gnysek/Input/pull/4/files

**Comment by gnysek** [comment]
updated (I have no option to re-open from what I see)

**Comment by JujuAdams** [comment]
There's a lot to like here! There's some extra behaviour I'd like to put in to tweak things, but this is in a very good position. I'll merge this in after 5.1.1 gets released (bugfixes for 5.1.0) and then have a play.

**Pull request merged**